In [2]:
import pandas as pd

In [5]:
df=pd.read_csv('GroceryStoreDataSet.csv',names=['Products'],header=None)  #here header none denotes that dont treat the first row as header name and store the data into col name Products

In [6]:
df.head(5)

Products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [7]:
data=list(df['Products'].apply(lambda x:x.split(',')))

In [10]:
type(df['Products'])

pandas.core.series.Series

In [11]:
df['Products'][0]

'MILK,BREAD,BISCUIT'

In [13]:
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [14]:
data[0]

['MILK', 'BREAD', 'BISCUIT']

In [16]:
type(data[0])

list

In [17]:
from mlxtend.preprocessing import TransactionEncoder

In [18]:
te=TransactionEncoder()

In [19]:
te_data=te.fit(data).transform(data)

In [25]:
te.columns_

['BISCUIT',
 'BOURNVITA',
 'BREAD',
 'COCK',
 'COFFEE',
 'CORNFLAKES',
 'JAM',
 'MAGGI',
 'MILK',
 'SUGER',
 'TEA']

In [26]:
df2=pd.DataFrame(te_data,columns=te.columns_)

In [27]:
df2

BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0      True      False   True  False   False       False  False  False   True   
1      True      False   True  False   False        True  False  False   True   
2     False       True   True  False   False       False  False  False  False   
3     False      False   True  False   False       False   True   True   True   
4      True      False  False  False   False       False  False   True  False   
5     False       True   True  False   False       False  False  False  False   
6     False      False  False  False   False        True  False   True  False   
7      True      False   True  False   False       False  False   True  False   
8     False      False   True  False   False       False   True   True  False   
9     False      False   True  False   False       False  False  False   True   
10     True      False  False   True    True        True  False  False  False   
11     True      False  False   True    True        True  False  False  False   
12    False       True  False  False    True       False  False  False  False   
13    False      False   True   True    True       False  False  False  False   
14     True      False   True  False   False       False  False  False  False   
15    False      False  False  False    True        True  False  False  False   
16    False       True   True  False   False       False  False  False  False   
17    False      False   True  False    True       False  False  False  False   
18    False      False   True  False    True       False  False  False  False   
19    False      False  False  False    True        True  False  False   True   

    SUGER    TEA  
0   False  False  
1   False  False  
2   False   True  
3   False  False  
4   False   True  
5   False   True  
6   False   True  
7   False   True  
8   False   True  
9   False  False  
10  False  False  
11  False  False  
12   True  False  
13  False  False  
14   True  False  
15   True  False  
16   True  False  
17   True  False  
18   True  False  
19  False   True

In [28]:
from mlxtend.frequent_patterns import apriori,association_rules
import numpy as np

In [29]:
df3=apriori(df2,min_support=0.15,use_colnames=True)

In [30]:
df3

support               itemsets
0      0.35              (BISCUIT)
1      0.20            (BOURNVITA)
2      0.65                (BREAD)
3      0.15                 (COCK)
4      0.40               (COFFEE)
5      0.30           (CORNFLAKES)
6      0.25                (MAGGI)
7      0.25                 (MILK)
8      0.30                (SUGER)
9      0.35                  (TEA)
10     0.20       (BISCUIT, BREAD)
11     0.15  (CORNFLAKES, BISCUIT)
12     0.15     (BOURNVITA, BREAD)
13     0.15        (COFFEE, BREAD)
14     0.15         (MAGGI, BREAD)
15     0.20          (MILK, BREAD)
16     0.20         (SUGER, BREAD)
17     0.20           (TEA, BREAD)
18     0.15         (COFFEE, COCK)
19     0.20   (CORNFLAKES, COFFEE)
20     0.20        (SUGER, COFFEE)
21     0.20           (MAGGI, TEA)

In [35]:
df3.sort_values(by='support',ascending=False)

support               itemsets
2      0.65                (BREAD)
4      0.40               (COFFEE)
0      0.35              (BISCUIT)
9      0.35                  (TEA)
5      0.30           (CORNFLAKES)
8      0.30                (SUGER)
6      0.25                (MAGGI)
7      0.25                 (MILK)
15     0.20          (MILK, BREAD)
20     0.20        (SUGER, COFFEE)
19     0.20   (CORNFLAKES, COFFEE)
17     0.20           (TEA, BREAD)
16     0.20         (SUGER, BREAD)
21     0.20           (MAGGI, TEA)
1      0.20            (BOURNVITA)
10     0.20       (BISCUIT, BREAD)
14     0.15         (MAGGI, BREAD)
13     0.15        (COFFEE, BREAD)
12     0.15     (BOURNVITA, BREAD)
18     0.15         (COFFEE, COCK)
3      0.15                 (COCK)
11     0.15  (CORNFLAKES, BISCUIT)

In [36]:
df4=association_rules(df3,metric='confidence',min_threshold=0.6)

In [37]:
df4

antecedents consequents  antecedent support  consequent support  support  \
0   (BOURNVITA)     (BREAD)                0.20                0.65     0.15   
1       (MAGGI)     (BREAD)                0.25                0.65     0.15   
2        (MILK)     (BREAD)                0.25                0.65     0.20   
3       (SUGER)     (BREAD)                0.30                0.65     0.20   
4        (COCK)    (COFFEE)                0.15                0.40     0.15   
5  (CORNFLAKES)    (COFFEE)                0.30                0.40     0.20   
6       (SUGER)    (COFFEE)                0.30                0.40     0.20   
7       (MAGGI)       (TEA)                0.25                0.35     0.20   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.750000  1.153846    0.0200       1.400       0.166667  
1    0.600000  0.923077   -0.0125       0.875      -0.100000  
2    0.800000  1.230769    0.0375       1.750       0.250000  
3    0.666667  1.025641    0.0050       1.050       0.035714  
4    1.000000  2.500000    0.0900         inf       0.705882  
5    0.666667  1.666667    0.0800       1.800       0.571429  
6    0.666667  1.666667    0.0800       1.800       0.571429  
7    0.800000  2.285714    0.1125       3.250       0.750000

In [38]:
filter=df4[(df4['support']>=0.15) & (df4['confidence']>=0.75)]

In [39]:
filter

antecedents consequents  antecedent support  consequent support  support  \
2      (MILK)     (BREAD)                0.25                0.65     0.20   
4      (COCK)    (COFFEE)                0.15                0.40     0.15   
7     (MAGGI)       (TEA)                0.25                0.35     0.20   

   confidence      lift  leverage  conviction  zhangs_metric  
2         0.8  1.230769    0.0375        1.75       0.250000  
4         1.0  2.500000    0.0900         inf       0.705882  
7         0.8  2.285714    0.1125        3.25       0.750000

#here antecedent support is 0.25 means milk is purchased 25% in all the transactions and consequent support is 0.65 means bread is purchased 65% in the transactions #now support indicates that milk and bread are observed together in 20% of the transaction and confidence 0.8 means if milk is bought bread is 80%likely to be bought as well 
#lift 1.23 means milk and bread affects each others sales positively. Sales of bread increases by 1.23 in transactions which contains milk or viceversa